src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs

```diff
@@ -27,42 +30,192 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
+        {
+            private enum SuppressFinalizeUsage
+            {
+                CanCall,
+                MustCall,
+                MustNotCall
+            }
+
+            private readonly Compilation _compilation;
+            private readonly IMethodSymbol _containingMethodSymbol;
+            private readonly IMethodSymbol _gcSuppressFinalizeMethodSymbol;
+            private readonly SuppressFinalizeUsage _expectedUsage;
+
+            private bool _suppressFinalizeCalled = false;
+            private SemanticModel _semanticModel;
+
+            public SuppressFinalizeAnalyzer(IMethodSymbol methodSymbol, IMethodSymbol gcSuppressFinalizeMethodSymbol, Compilation compilation)
+            {
+                this._compilation = compilation;
+                this._containingMethodSymbol = methodSymbol;
+                this._gcSuppressFinalizeMethodSymbol = gcSuppressFinalizeMethodSymbol;
+
+                this._expectedUsage = GetAllowedSuppressFinalizeUsage(_containingMethodSymbol);
+            }
+
+            public void Analyze(OperationAnalysisContext analysisContext)
+            {
+                var invocationExpression = (IInvocationExpression)analysisContext.Operation;
+                if (invocationExpression.TargetMethod.OriginalDefinition.Equals(_gcSuppressFinalizeMethodSymbol))
+                {
+                    _suppressFinalizeCalled = true;
+
+                    if (_semanticModel == null)
+                    {
+                        _semanticModel = analysisContext.Compilation.GetSemanticModel(analysisContext.Operation.Syntax.SyntaxTree);
+                    }
+
+                    // Check for GC.SuppressFinalize outside of IDisposable.Dispose()
+                    if (_expectedUsage == SuppressFinalizeUsage.MustNotCall)
+                    {
+                        analysisContext.ReportDiagnostic(invocationExpression.Syntax.CreateDiagnostic(
+                            OutsideDisposeRule,
+                            _containingMethodSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat),
+                            _gcSuppressFinalizeMethodSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat)));
+                    }
+
+                    // Checks for GC.SuppressFinalize(this)
+                    if (invocationExpression.ArgumentsInSourceOrder.Count() != 1)
+                    {
+                        return;
+                    }
+
+                    var parameterSymbol = _semanticModel.GetSymbolInfo(invocationExpression.ArgumentsInSourceOrder.Single().Syntax).Symbol as IParameterSymbol;
+                    if (parameterSymbol == null || !parameterSymbol.IsThis)
+                    {
+                        analysisContext.ReportDiagnostic(invocationExpression.Syntax.CreateDiagnostic(
+                            NotPassedThisRule,
+                            _containingMethodSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat),
```

Comment:
Most often people view the error list when they have context and so short names tend to be enough to disambiguate. Fully qualified names tend to make the message much longer and make the error unintelligible.